Update scopes docs (#7071)
auth0/docs: articles/scopes/current/api-scopes.md

---
description: Understand scopes used with APIs.
topics:
  - scopes
contentType:
  - concept
useCase:
  - development
  - call-api
  - secure-api
---
# API Scopes

As an [API](/apis) developer, you need to define the scopes available for applications that might call your API. This way, you can apply fine-grained control to the information and actions available to your users. In this case, you need to define custom scopes for your API and then identify these scopes so that calling applications can use them.

For example, let's say you are building an API that provides data to a calendar application. You want some users to be able to edit items on the calendar, others to only be able to read them, and others to be able to both read and write to calendar items. To do this, you create two scopes for your API: one that authorizes write access (`write:appointments`) and one that authorizes read-only access (`read:appointments`). 

> Now, when an app calls your API, it will specify the scope it needs in its request. The app may request read access by including `read:appointments` in its scope, write access by including `write:appointments` in its scope, or both read and write access by including both `read:appointments` and `write:appointments` in its scope.

Reviewer: So, each of those two items are scopes; so it's not that the app is including those items in its scope, but rather its request includes the `write:appointments` and `read:appointments` scopes.

For an example showing how to request custom API access for your application, see [Sample Use Cases: Scopes and Claims](/scopes/current/sample-use-cases#request-custom-API-access).

## Limit API scopes

> An application can request any defined scope and the user will be prompted to approve that scope during the authorization flow. Instead, you may want to limit available scopes based on, for example, the role of a user. To do this, you can use the [Authorization Extension](/extensions/authorization-extension) and a custom [Rule](/rules).

Reviewer: This seems to imply that you would be limiting the scopes available to an application based on roles, which isn't quite accurate.

Maybe something like this: `An application can include any defined scope in its request. Instead of including _all_ available scopes in every request, however, the scopes can be limited. For example, a user of your application can be given a role so that requests on their behalf are limited to just the scopes assigned to that role. To do this, you...`

---
description: Understand permissions (scopes) used with APIs.
topics:
  - scopes
  - permissions
contentType:
  - concept
useCase:
  - development
  - call-api
  - secure-api
---
# API Permissions

As an [API](/apis) developer, you need to:

> 1. Decide which user information you would like external applications to be able to access on a user's behalf.

Reviewer 2: Calling this "user information" makes it sounds like its user profile information. Rather, this is data (information) within the API that can be _managed by_ or is _associated with_ the user.

```suggestion
1. Decide which information you would like external applications to be able to access on a user's behalf.
```

> 2. Define these access levels as custom permission levels.

Reviewer 2: Using the term "permission levels" is confusing as it implies some variation of the concept of a permission. I would just use "permissions".

```suggestion
2. Define these access levels as custom permissions.
```

> 3. Identify these permission levels so that calling applications can use them. 

Reviewer 2: Same reasoning as the previous comment:

```suggestion
3. Identify these permissions so that calling applications can use them. 
```

For example, let's say you are building an API that provides data to a calendar application. At various times, the app may need to edit items on the calendar, read them, or both read and write to calendar items for a user. To do this, you create two permission levels for your API: one that authorizes write access (`write:appointments`) and one that authorizes read-only access (`read:appointments`). 

> A calling application will request authorization from the user to access the requested permissions, and the user will approve or deny the request. The app may request read access by including the `read:appointments` permission in its request, write access by including the `write:appointments` permission in its request, or both read and write access by including both the `read:appointments` and `write:appointments` permissions in its request. 

Reviewer 2: I've already raised this issue in the main docs page (`/scopes/current`), but this language:
>A calling application will request authorization from the user to access the requested permissions

is again why I think we need to separate the concepts of "permissions" and "scopes". Permissions are about what the user themselves has been authorized to do. What we're talking about here is delegated access, which is what scopes are for. I personally think this document should be under the "scopes" umbrella, assuming we break out both concepts.

> Now, when the app calls your API, it will include a token which verifies that the user has provided authorization to access their content and also indicates which permissions the user has approved. Your API should respect the approved permissions and only release information that was authorized by the user to the calling application.

Reviewer 3: I think we should add here that an API can use additional logic beyond the token itself to allow access or not. APIs may choose to enforce additional access control mechanisms.

Author: Added note.

For an example showing how to request custom API access for your application, see [Sample Use Cases: Permissions and Claims](/scopes/current/sample-use-cases#request-custom-API-access).

> ## Limit API permissions

Reviewer 2: This section attempts to blur the lines a bit between user permissions and scopes. If you attempt to marry the two concepts, then it makes sense that a user can only delegate access to the things they themselves have been given permission to access. While I think this is a valid simplification in some use cases, I think it breaks down in more complex ones. Again, I think we need to keep these concepts separate. In other words, an API may have both scopes and permissions.

> An application can include any defined permission level in its request. Instead of including _all_ available permissions in every request, however, the permissions can be limited. For example, a user of your application can be given a role so that requests on their behalf are limited to just the permissions assigned to that role. To do this, you can use the [Authorization Extension](/extensions/authorization-extension) and a custom [Rule](/rules).

Reviewer 3: I don't follow this block. Are we simply trying to say an app should limit its request only to what it needs?
